MemMrg/memmrg.cpp

```diff
@@ -0,0 +1,64 @@
+#include "memmrg.h"
+
+#include <stdexcept>
+
+const int MemMrg::item_size_(32);
+
+
+MemMrg::MemMrg(const int max_size):
+    item_cnt_(max_size),
+    mrg_buf_(NULL),
+    head_(NULL)
+{
+    mrg_buf_ = new char[item_size_*item_cnt_];;
+    head_ = reinterpret_cast<void*> (mrg_buf_);
+
+
+    for (int i = 0; i < item_size_*(item_cnt_-1); i+=item_size_) {
+        char **curr = reinterpret_cast<char**> (mrg_buf_+i);
+        *curr = mrg_buf_ + i + item_size_;
+    }
+
+    char **last = reinterpret_cast<char**> (mrg_buf_ + item_size_*(item_cnt_-1));
+    *last = NULL;
+}
+
+MemMrg::~MemMrg()
+{
+    delete [] mrg_buf_;
+}
+
+void *MemMrg::my_alloc()
+{
+    //no free memory in buffer
+    if (head_ == NULL)
+        throw std::bad_alloc();
+
+    void *result = head_;
+    head_ = *reinterpret_cast<void**> (head_);
```

Comment:
It's all Ok with pointers arifmetics, but why do we write it as char** and reed as void**?